### PR TITLE
docs(factories): document custom webhooks

### PR DESCRIPTION
**Pull request by advait-m**
## What this feature does

Custom webhooks give a factory an authenticated HTTPS URL that any system can POST JSON to (internal CI, PagerDuty, Sentry, Stripe, homegrown tools). Automations subscribe to a webhook, filter deliveries with a declarative pattern over the payload, and start runs with the payload attached. Three authentication modes: bearer token, URL token, and provider signature. Shipped server-side in the warp-server `release/2026-09-01` cut (REMOTE-2447 stack plus the REMOTE-2956 dashboard UI).

## Summary

Adds a Webhooks page to the Factories tab and threads custom webhooks through the pages that already describe how work reaches a factory. Written against `warp-server` `develop` (specs `specs/REMOTE-2447/`, PRs #13771, #13831, #13833, #13915, #13916, #14009, #15686, #15722, #16365, #16417) and the factory dashboard client in `client/packages/factory`.

The payload-pattern grammar lives on the Automations page (the canonical home for filter matching rules) and the `webhooks/<name>.yaml` key reference lives on the Definitions as code page, so the webhooks page stays focused on the setup flow and links out for both.

### src/content/docs/factories/webhooks.mdx
- Restructured around the setup flow: what a webhook does → authentication modes → create the webhook → configure the sender → add a **Webhook** trigger to an automation → filter deliveries by payload → what the run receives
- Exact HTTP semantics, delivery identity/dedup, ordering, and limits now live in a **Delivery rules** reference section further down the page (after Manage webhooks), not the conceptual intro
- Secret rotation/rollover detail moved into **Manage webhooks**; dropped the "URL is safe to share" framing from Authentication modes
- The create-webhook flow's default path is now "keep the suggested secret name," with the managed-secret / reuse-an-existing-secret detail moved to an advisory note

### src/content/docs/factories/automations.mdx
- Custom webhooks row in the "What each source can filter on" table
- "Payload filters for webhook triggers" simplified to the mental model, supported operators, and the GitHub example + dashboard testing flow
- New **Payload filter reference** section holds numeric equality semantics, missing-key behavior, array-of-object matching, nesting/leaf/value limits, conflicting-operator validation, and fail-closed behavior
- Access note for webhook-authenticated senders, a troubleshooting bullet, and a related-pages link

### src/content/docs/factories/factory-as-code.mdx
- `webhook` / `received` added to the trigger provider list, plus `webhook_ids` and `payload` under `triggers[].filter`
- New `webhooks/<name>.yaml` section documenting `authMode`, `signatureScheme`, `secretName`, `deliveryIdHeader`, and `enabled`
- Directory tree and description updated

### Cross-references
- `factories/connect-your-factory.mdx` - Custom webhooks row in the source table and a short section
- `factories/factory-dashboard.mdx` - **Webhooks** page mentioned alongside Agents and Automations
- `factories/troubleshooting.mdx` - Custom webhooks row in the "work isn't starting" source table
- `factories/index.mdx` - custom webhooks listed among work sources
- `platform/triggers/index.mdx` - Custom webhooks bullet under available trigger types
- `src/sidebar.ts` - Webhooks entry in the Factories tab's Integrations group, between Jira and Factory API

## Content design plan

**Audience and JTBD:** A platform or infra engineer who already runs a factory and wants an event from a tool Warp doesn't integrate with (their CI, PagerDuty, Sentry, Stripe, an internal service) to start factory work, without writing a bot against the factory API.

**Problem:** Today the docs list Slack, GitHub, GitLab, Linear, Jira, the factory API, and Factory MCP as the only ways in. The **Webhooks** page and **Webhook** trigger now exist in the dashboard with no docs behind them, so a reader can't tell which authentication mode their sender supports, why a delivery returned `202` but started nothing, or how to write a payload filter that isn't the flat key/value shape the other filters use.

**Goals:**
- The reader can pick the authentication mode their sender supports, create the webhook, and confirm a test delivery lands.
- The reader can bind an automation to the webhook, filter on the payload, and verify the filter against a real delivery before saving.
- The reader can diagnose a delivery that didn't start a run from the response code and the delivery log.

**Purpose and value:** Custom webhooks are the escape hatch for every tool without a first-party integration. Without this page, the only documentation is the dashboard's inline hints, and the payload grammar (operators, nesting, missing-key semantics) is not discoverable at all.

**Content type:** Feature documentation (combined) — the reader needs both the concept (what a delivery is, how identity and dedupe work) and the task (create, configure the sender, add the trigger) in one place.

**Skill and template:** `draft_feature_doc` / `.agents/templates/feature-doc.md`

**High-impact scenarios:**
- Covers: bearer-token sender from CI or a script; a signed provider (Sentry example); binding an automation with a payload filter and testing it; rotation, disable, and delete; defining webhooks in a GitHub-backed definition.
- Excludes: the internal `/factory/webhooks` REST endpoints — they are `x-internal` in the OpenAPI spec and not part of the public API reference. Per-provider sender configuration screens (where the Stripe or PagerDuty webhook form lives) — that's each provider's own docs. Delivery match counts — the column exists but isn't populated yet, so it isn't documented.

## Unverified claims

None — UI labels (**Webhooks**, **Add webhook**, "Authentication", "Provider scheme", "Provider secret", "Secret name", "Delivery ID header", **Create webhook**, **Add trigger** > **Webhook**, "Webhook received from", **Payload filter**, **Test filter**, **Would trigger** / **Would not trigger**, **Recent deliveries**, **Rotate secret**, delivery outcome labels), response codes, limits (20 webhooks per factory, 256 KB body, 60 deliveries/min, 50-delivery / 7-day retention), delivery-ID precedence, the 5-minute signature tolerance, the ingress URL host (`webhook_origin` in `config/prod.yaml`), and the `webhooks/<name>.yaml` schema were all verified against `warp-server` `develop` (`06c178828a`).

### Review changes

Addressed Hong Yi's review (restructure request plus 7 inline comments):
- Restructured `webhooks.mdx` around the task flow per the review body's suggested arc (what a webhook does → create → configure sender → add to automation → filter → test), pushing HTTP semantics, dedup, secret rotation, retention limits, and ordering into reference sections lower on the same page (kept as one dedicated page per the requester's call, not split into `automations.mdx`/`factory-as-code.mdx`)
- Simplified "How custom webhooks work" to the user mental model; moved `202`/delivery-ID detail into Delivery rules
- Cut the ordering/no-threading note from the conceptual intro; kept a one-line reference bullet instead of a callout, since it's backend detail rather than something users decide around
- Kept Authentication modes focused on choosing a mode; moved secret rotation/rollover into Manage webhooks; dropped the "URL is safe to share" framing entirely
- Changed the create-webhook default path to "keep the suggested secret name"; moved the managed-secret / reuse-existing-secret detail into an advisory note
- Simplified the payload filter section in `automations.mdx` to the mental model, operators, and existing example; moved numeric equality, missing-key behavior, array-of-object matching, limits, validation, and fail-closed behavior into a new Payload filter reference section
- Recaptured both screenshots: the Webhook created pane no longer includes the dimmed background list, and the trigger-row screenshot no longer clips the "Test against" delivery label
- Replied on each of the 7 threads with specifics and resolved them; followed up on the earlier page-length thread noting the restructure moved Delivery rules out of the happy path too

Verification re-run after this pass: `check_for_broken_links.py --internal-only` (0 broken), `style_lint.py --changed` (0 errors), `npm run build` (succeeds).

## Notes for reviewers

- Screenshots: Factory MSW captures of the Webhook created secret-reveal pane and the Webhook trigger row with Payload filter expanded and a successful **Test filter**. All URLs, UIDs, delivery IDs, and secrets in the captures are fake (`example.test`, `whsec_docs_example_only`, `dlv_docs_example`).
- `webhook_ids` in definition files takes UIDs only (name aliases are a listed follow-up in the spec), so the page tells GitHub-backed readers to sync the webhook file first, then copy the UID from the dashboard. Worth a sanity check that this is the flow we want to document rather than wait on aliases.
- `trunk check` wasn't run (Trunk isn't installed locally). `style_lint --changed` reports no errors, the internal link checker passes, and `npm run build` succeeds.

[Warp conversation](https://staging.warp.dev/conversation/3e023374-b5e1-43c6-add8-ef10d414d641)
Originating thread: https://warpdev.slack.com/archives/C09BVK0PL3Y/p1788310815659089

Co-Authored-By: Warp <agent@warp.dev>

<details>
<summary>Computer-use screenshots (2)</summary>

Narrow 520x430 viewport screenshot of the 'Webhook created' panel showing the yellow one-time-secret warning, Ingress URL (https://example.test/webhooks/w/wh\_docs\_created\_example), Secret (whsec\_docs\_example\_only), and Done button, with a thin sliver of the dimmed webhook list visible on the left.
Narrow 520x430 viewport screenshot of the 'Webhook created' panel showing the yellow one-time-secret warning, Ingress URL (https://example.test/webhooks/w/wh\_docs\_created\_example), Secret (whsec\_docs\_example\_only), and Done button, with a thin sliver of the dimmed webhook list visible on the left.

Final tightly-framed screenshot (848×351px) of the webhook trigger section showing "Webhook received from Internal CI", the expanded Payload filter with JSON content, the "Test against dlv\_docs\_example · 2 minutes ago" row, "Test filter" button, and the green "Would trigger" badge with "Source matched · Payload matched" text — nothing clipped at any edge.
Final tightly-framed screenshot (848×351px) of the webhook trigger section showing "Webhook received from Internal CI", the expanded Payload filter with JSON content, the "Test against dlv\_docs\_example · 2 minutes ago" row, "Test filter" button, and the green "Would trigger" badge with "Source matched · Payload matched" text — nothing clipped at any edge.

</details>

### CODE REVIEW

File: src/content/docs/factories/webhooks.mdx [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+The pane shows the ingress URL and the secret; for a URL token webhook, the secret is embedded in the URL itself. Copy them now: Warp doesn't show the secret again, and a URL token webhook's URL renders a `{token}` placeholder on every later view. If you referenced an existing secret, the value isn't shown at all; configure the sender with the value you stored in that secret.
+
+<figure style={{ maxWidth: "375px" }}>
+![The Webhook created pane showing a one-time ingress URL on example.test and a secret that starts with whsec_.](../../../assets/factories/webhook-created-secret-reveal.png)
+<figcaption>The one-time URL and secret after creating a webhook.</figcaption>
+</figure>
+
+### Configure the sender
+
+Point the sender at the ingress URL and give it the credential for the webhook's mode. To confirm the webhook works before wiring up the real sender, post a test delivery:
```
[the remark this answers is not in the document]
Comment (advait-m) [reply src/content/docs/factories/webhooks.mdx]: [Warp Agent] Agreed. The example only authenticated the bearer token mode. Reworked "Configure the sender" in https://github.com/warpdotdev/docs/pull/671/commits/8ca1fb07a3bddb1669cca740c0210c39f0204fcd: the curl is now explicitly the bearer token path, URL token readers are told to drop the `Authorization` header and post to the full copied URL (which already carries the token), and provider signature readers are pointed at the provider's own test-event feature with a note that an unsigned request returns `401`.

File: src/content/docs/factories/webhooks.mdx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+```bash
+curl -X POST "https://app.warp.dev/webhooks/w/WEBHOOK_UID" \
+  -H "Authorization: Bearer WEBHOOK_SECRET" \
+  -H "Content-Type: application/json" \
+  -d '{"event": "deploy.finished", "env": "production", "service": "payments"}'
+```
+
+Replace `WEBHOOK_UID` with the UID in the ingress URL and `WEBHOOK_SECRET` with the secret you copied. The response is `202` with a `delivery_id`, and the delivery appears under **Recent deliveries** when you open the webhook on the **Webhooks** page.
```
[the remark this answers is not in the document]
Comment (advait-m) [reply src/content/docs/factories/webhooks.mdx]: [Warp Agent] Fixed in https://github.com/warpdotdev/docs/pull/671/commits/8ca1fb07a3bddb1669cca740c0210c39f0204fcd, along the lines you suggested. "Configure the sender" now labels the curl as the bearer token path, then covers the other two: a URL token webhook posts to the full copied URL with no `Authorization` header, and a provider signature webhook is exercised from the provider's own test-event feature (an unsigned request returns `401`). The "accepted delivery returns `202` and appears under Recent deliveries" outcome now reads as mode-agnostic.

File: src/content/docs/factories/webhooks.mdx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+* **Rotate secret** - Generates a new secret, or takes the provider's new signing secret for a provider signature webhook, and shows it once. The old secret stops working immediately.
+* **Delete** - Senders get `404` immediately. Automations that select the deleted webhook stay visible but stop firing; point their triggers at another webhook or remove the trigger.
+
+On a Warp-managed factory, the dashboard writes each webhook to a definition file, so two actions work differently: rename isn't supported, and **Rotate secret** is replaced by rotating the managed secret the webhook references on the team's **Secrets** page. The webhook picks up the new value on its next sync.
+
+## Webhooks in definitions as code
+
+In a [factory definition](/factories/factory-as-code/), each webhook is a `webhooks/<name>.yaml` file, and an automation subscribes with a `webhook` trigger. The file never holds the secret: `secretName` references a managed secret that already exists on the team, in every authentication mode.
```
[the remark this answers is not in the document]
Comment (advait-m) [reply src/content/docs/factories/webhooks.mdx]: [Warp Agent] Partially taken in https://github.com/warpdotdev/docs/pull/671/commits/8ca1fb07a3bddb1669cca740c0210c39f0204fcd. Two notes on the measurement first: the ~2,100 figure counts frontmatter and code blocks; the prose was ~1,870 words and is about the same after this pass, because the fix for the test-delivery comment above added ~60 words.

What changed: the definitions-as-code section no longer restates the `secretName` rules or the URL-safe-character constraint (both now live only under `webhooks/<name>.yaml` on the definitions reference), and the example is trimmed. I kept one compact example rather than reducing the section to a link: the webhook file and its `webhook` trigger side by side, plus the "UIDs not file names, sync first, then copy the UID" workflow, aren't shown together anywhere else and are the non-obvious part for GitHub-backed factories.

The remaining length is the delivery contract and the auth-mode table, which this page owns. If you'd rather keep the feature page under budget, "Delivery rules" could move to a reference page; leaving this thread open for @advait-m to call.

File: src/content/docs/factories/factory-as-code.mdx [hunk 4]
```diff
@@ -267,14 +269,27 @@ The providers and their events:
...
 
 ### `triggers[].filter`
 
 Optional. Narrows which events start runs. The keys a filter accepts depend on the provider and event: for example `repos`, `labels`, and `authors` for GitHub events, or `channels`, `users`, and `keywords` for Slack messages. Filter keys combine with AND, an omitted key matches everything, and each key takes a list that matches any of its values (or an `in`/`not_in` object to include or exclude). Slack and Linear filters take names (channels, users, teams, projects, states), and Warp resolves them to IDs when it applies the change.
 
+A `webhook` / `received` trigger's filter requires `webhook_ids`, a list of webhook UIDs (`in` only; the file name isn't accepted), and takes an optional `payload` pattern that mirrors the delivery's JSON body. See [payload filters for webhook triggers](/factories/automations/#payload-filters-for-webhook-triggers) for the pattern grammar.
```
[the remark this answers is not in the document]
Comment (advait-m) [reply src/content/docs/factories/factory-as-code.mdx]: [Warp Agent] Applied your suggestion verbatim in https://github.com/warpdotdev/docs/pull/671/commits/8ca1fb07a3bddb1669cca740c0210c39f0204fcd. It was the only slashed shorthand in the diff.

File: src/assets/factories/webhook-created-secret-reveal.png
Comment (hongyi-chen) [inline src/assets/factories/webhook-created-secret-reveal.png:1]: could we crop this a bit better somehow, how about this? 

Image

File: src/assets/factories/webhook-payload-filter-test.png
Comment (hongyi-chen) [inline src/assets/factories/webhook-payload-filter-test.png:1]: similarly, i feel like there is some content that is being cut off here

File: src/content/docs/factories/automations.mdx [hunk 5]
```diff
@@ -92,14 +93,47 @@ the PR.
 
 For more automations that use these filters, including branches, paths, workflow conclusions, and emoji reactions, see [`06-common-automations`](https://github.com/warpdotdev/warp-factory-examples/tree/main/examples/06-common-automations).
 
+## Payload filters for webhook triggers
```
Comment (hongyi-chen) [inline src/content/docs/factories/automations.mdx:96]: I think this is overexplaining the filter implementation a bit for the main docs. As a user, I mostly need to understand the mental model, see the supported operators, and get a couple examples.

Could we simplify this to something like:

* payload filters match against the webhook JSON body
* fields are ANDed together
* arrays mean “match any of these values”
* in, not_in, and exists are supported
* nested objects let you filter on nested payload fields

Then keep the GitHub example + dashboard testing flow. Details like numeric equality semantics, missing-key behavior, array-of-object matching, nesting/leaf/value limits, conflicting operator validation, and failure behavior feel more like reference/implementation details that could live in a separate “filter reference” section if we think users need them.

File: src/content/docs/factories/webhooks.mdx [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+```mermaid
+flowchart LR
+  Sender["External system"] -->|"POST JSON"| URL["Webhook URL"]
+  URL --> Log["Delivery log"]
+  URL --> Match["Automations with a<br/>matching Webhook trigger"]
+  Match --> Run["Factory run with the<br/>payload attached"]
+```
+
+When a delivery arrives, Warp authenticates it, records it in the webhook's delivery log, and returns `202 Accepted` with a delivery ID. Every automation whose **Webhook** trigger selects that webhook and whose payload filter matches starts one run, with the JSON body attached as `event-payload.json`. A `202` means the delivery was accepted for evaluation, not that an automation fired: a delivery that matches nothing is still accepted and logged.
```
Comment (hongyi-chen) [inline src/content/docs/factories/webhooks.mdx]: I think we can simplify this section around the user mental model rather than describing the processing sequence. Something like: “When an external system sends an event to your webhook, Warp checks whether it matches any automation triggers and starts the corresponding runs.”

The 202 semantics / delivery ID behavior feels more appropriate for Delivery rules or troubleshooting.

File: src/content/docs/factories/webhooks.mdx [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+Deliveries are independent: there's no ordering guarantee between them and no reply or thread continuation. Each accepted delivery starts new runs or nothing.
```
Comment (hongyi-chen) [inline src/content/docs/factories/webhooks.mdx]: Do we expect users to make decisions based on this? If ordering / threading is an important product limitation, I’d frame it as a short callout. Otherwise this feels like backend semantics we don’t need in the conceptual intro.

File: src/content/docs/factories/webhooks.mdx [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+## Authentication modes
```
Comment (hongyi-chen) [inline src/content/docs/factories/webhooks.mdx:25]: The table gives me basically everything I need here. Could we keep this section focused on choosing an auth mode and move secret lifecycle / rollover details into Manage webhooks or a security/reference callout?

“URL is safe to share” also makes me a little nervous as user-facing guidance. Even if technically true, I’m not sure we need to encourage users to think about the URL that way.

File: src/content/docs/factories/webhooks.mdx [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+### Prerequisites
+
+* **Permission to manage the factory** - Creating and editing webhooks changes the factory's configuration.
+
+### Create the webhook
+
+1. In the factory dashboard, open **Webhooks** and click **Add webhook**.
+2. Enter a "Name" the automation editor will show, such as `Internal CI`.
+3. In the "Authentication" dropdown, choose **Bearer token**, **URL token**, or **Provider signature**. For a provider signature, also choose the "Provider scheme" and paste the "Provider secret" from the provider.
+4. On a Warp-managed factory, the "Secret name" field names the [managed secret](/platform/secrets/) that stores this webhook's secret. Keep the suggested name to have Warp create the secret, or enter the name of an existing team secret to use its value instead.
```
Comment (hongyi-chen) [inline src/content/docs/factories/webhooks.mdx]: This feels like we’re exposing the underlying configuration model in the happy-path setup flow. Does a user actually need to understand that the webhook secret is backed by a managed secret?

If the UI already suggests a name, I’d make the default path simply “Keep the suggested secret name,” and move reuse-of-existing-secrets into an advanced note.

**Review by hongyi-chen — COMMENTED** [review]
Overall I think the content is comprehensive, but the main guide is doing three jobs at once: onboarding, conceptual documentation, and protocol reference.

I’d optimize this page for “I want to connect something to my Factory with a webhook”:

1. what a webhook does
2. create one
3. configure the sender
4. add it to an automation
5. optionally filter the payload
6. test it

Then keep exact HTTP semantics, deduplication rules, secret implementation, filter edge cases, retention limits, and definition schema behavior in reference/troubleshooting sections. Right now we expose a lot of the machinery before the user has a reason to care about it.

**Review by hongyi-chen — APPROVED** [review]
